core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,270 @@
+            this.browserUrl = browserUrl;
+            this.authenticationName = authenticationName;
+        }
+
+        public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException {
+            String restOfPath = req.getRestOfPath();
+
+            if (restOfPath.isEmpty()) {
+                // Do not go through resource URLs for empty rest-of-path
+                String url = Jenkins.get().getRootUrl() + browserUrl;
+                rsp.sendRedirect(302, url);
+                return;
+            }
+
+            AccessControlled requestRoot = Jenkins.get();
+            String requestUrlSuffix = this.browserUrl;
+
+            if (LOGGER.isLoggable(Level.FINE)) {
+                LOGGER.log(Level.FINE, "Performing a request as authentication: " + authenticationName + " to object: " + requestRoot + " and restOfUrl: " + requestUrlSuffix + " and restOfPath: " + restOfPath);
+            }
+
+            Authentication auth = Jenkins.ANONYMOUS;
+            if (authenticationName != null) {
+                User user = User.getById(authenticationName, false);
+                if (user != null) {
+                    auth = user.impersonate();
+                }
+            }
+
+            try (ACLContext ignored = ACL.as(auth)) {
+                Stapler.getCurrent().invoke(req, rsp, requestRoot, requestUrlSuffix + restOfPath);
+                /*
+                TODO I am unsure what to do when this throws an exception.
+                Right now (below) we return 403/404 error pages. A more user-friendly approach is probably to
+                just redirect like we do for expired resource URLs, but the question is whether we'd end up in a
+                redirect loop if the exception is specific to this mode (and the "normal" URLs redirect to resource
+                URLs). That seems even worse.
+                 */
+            } catch (AccessDeniedException ade) {
+                LOGGER.log(Level.INFO, "Failed permission check for resource URL access", ade);
+                rsp.sendError(403, "Failed permission check: " + ade.getMessage());
+            } catch (Exception e) {
+                LOGGER.log(Level.INFO, "Something else failed for resource URL access", e);
+                rsp.sendError(404, "Failed: " + e.getMessage());
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "[" + super.toString() + ", authentication=" + authenticationName + "; key=" + browserUrl + "]";
+        }
+    }
+
+    public static class Token {
+        private String path;
+        private String username;
+        private long timestamp;
+        private Token (String path, @Nullable String username, long timestamp) {
+            this.path = path;
+            this.username = username;
+            this.timestamp = timestamp;
+        }
+
+        private String encode() {
+            String value = username + ":" + timestamp + ":" + path;
```

Comment:
This will give different values for `null` and `""` for `username` which you marked `@Nullable`.